Kubenet with IPv6 causes kubelet panic

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG REPORT

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature


**What happened**:
Using Kubenet with IPv6 causes a panic

panic: runtime error: index out of range

goroutine 625 [running]:
k8s.io/kubernetes/pkg/kubelet/network/kubenet.(*kubenetNetworkPlugin).Event(0xc4203bf440, 0x462ebd1, 0xf, 0xc4219deae0)
/root/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/network/kubenet/kubenet_linux.go:260 +0x889

This is due to IPv4 specific code in kubenet/kubenet_linux.go

**What you expected to happen**:
Kubenet should support IPv6

**How to reproduce it (as minimally and precisely as possible)**:

Running local-up-cluster.sh, use kubenet, and set the parameters
for kubenet with ipv6
node_cidr_args="--allocate-node-cidrs=true --cluster-cidr=2001:beef::/48 --node-cidr-mask-size=64 "

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
latest as of June 26, 2017
- Cloud provider or hardware configuration**:
local on a VM
- OS (e.g. from /etc/os-release):
CentOS
- Kernel (e.g. `uname -a`):
Linux localhost.localdomain 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Others:

--- a/pkg/kubelet/network/kubenet/kubenet_linux.go
+++ b/pkg/kubelet/network/kubenet/kubenet_linux.go
@@ -259,7 +259,7 @@ func (plugin *kubenetNetworkPlugin) Event(name string, details map[string]interf
 	if err == nil {
 		setHairpin := plugin.hairpinMode == componentconfig.HairpinVeth
 		// Set bridge address to first address in IPNet
-		cidr.IP.To4()[3] += 1
+		cidr.IP[len(cidr.IP)-1] += 1
 
 		json := fmt.Sprintf(NET_CONFIG_TEMPLATE, BridgeName, plugin.mtu, network.DefaultInterfaceName, setHairpin, podCIDR, cidr.IP.String())
 		glog.V(2).Infof("CNI network config set to %v", json)

--- a/pkg/kubelet/network/kubenet/kubenet_linux_test.go
+++ b/pkg/kubelet/network/kubenet/kubenet_linux_test.go
@@ -232,37 +232,61 @@ func TestGenerateMacAddress(t *testing.T) {
 // TestInvocationWithoutRuntime invokes the plugin without a runtime.
 // This is how kubenet is invoked from the cri.
 func TestTearDownWithoutRuntime(t *testing.T) {
-	fhost := nettest.NewFakeHost(nil)
-	fhost.Legacy = false
-	fhost.Runtime = nil
-	mockcni := &mock_cni.MockCNI{}
-
-	fexec := &exec.FakeExec{
-		CommandScript: []exec.FakeCommandAction{},
-		LookPathFunc: func(file string) (string, error) {
-			return fmt.Sprintf("/fake-bin/%s", file), nil
+	testCases := []struct {
+		podCIDR         string
+		ip              string
+		expectedGateway string
+	}{
+		{
+			podCIDR:         "10.0.0.1/24",
+			ip:              "10.0.0.1",
+			expectedGateway: "10.0.0.1",
+		},
+		{
+			podCIDR:         "2001:beef::1/48",
+			ip:              "2001:beef::1",
+			expectedGateway: "2001:beef::1",
 		},
 	}
+	for _, tc := range testCases {
+		fhost := nettest.NewFakeHost(nil)
+		fhost.Legacy = false
+		fhost.Runtime = nil
+		mockcni := &mock_cni.MockCNI{}
 
-	kubenet := newFakeKubenetPlugin(map[kubecontainer.ContainerID]string{}, fexec, fhost)
-	kubenet.cniConfig = mockcni
-	kubenet.iptables = ipttest.NewFake()
+		fexec := &exec.FakeExec{
+			CommandScript: []exec.FakeCommandAction{},
+			LookPathFunc: func(file string) (string, error) {
+				return fmt.Sprintf("/fake-bin/%s", file), nil
+			},
+		}
 
-	details := make(map[string]interface{})
-	details[network.NET_PLUGIN_EVENT_POD_CIDR_CHANGE_DETAIL_CIDR] = "10.0.0.1/24"
-	kubenet.Event(network.NET_PLUGIN_EVENT_POD_CIDR_CHANGE, details)
+		kubenet := newFakeKubenetPlugin(map[kubecontainer.ContainerID]string{}, fexec, fhost)
+		kubenet.cniConfig = mockcni
+		kubenet.iptables = ipttest.NewFake()
 
-	existingContainerID := kubecontainer.BuildContainerID("docker", "123")
-	kubenet.podIPs[existingContainerID] = "10.0.0.1"
+		details := make(map[string]interface{})
+		details[network.NET_PLUGIN_EVENT_POD_CIDR_CHANGE_DETAIL_CIDR] = tc.podCIDR
+		kubenet.Event(network.NET_PLUGIN_EVENT_POD_CIDR_CHANGE, details)
 
-	mockcni.On("DelNetwork", mock.AnythingOfType("*libcni.NetworkConfig"), mock.AnythingOfType("*libcni.RuntimeConf")).Return(nil)
+		if kubenet.gateway.String() != tc.expectedGateway {
+			t.Errorf("generated gateway: %q, expecting: %q", kubenet.gateway.String(), tc.expectedGateway)
+		}
+		if kubenet.podCidr != tc.podCIDR {
+			t.Errorf("generated podCidr: %q, expecting: %q", kubenet.podCidr, tc.podCIDR)
+		}
+		existingContainerID := kubecontainer.BuildContainerID("docker", "123")
+		kubenet.podIPs[existingContainerID] = tc.ip
 
-	if err := kubenet.TearDownPod("namespace", "name", existingContainerID); err != nil {
-		t.Fatalf("Unexpected error in TearDownPod: %v", err)
+		mockcni.On("DelNetwork", mock.AnythingOfType("*libcni.NetworkConfig"), mock.AnythingOfType("*libcni.RuntimeConf")).Return(nil)
+
+		if err := kubenet.TearDownPod("namespace", "name", existingContainerID); err != nil {
+			t.Fatalf("Unexpected error in TearDownPod: %v", err)
+		}
+		// Assert that the CNI DelNetwork made it through and we didn't crash
+		// without a runtime.
+		mockcni.AssertExpectations(t)
 	}
-	// Assert that the CNI DelNetwork made it through and we didn't crash
-	// without a runtime.
-	mockcni.AssertExpectations(t)
 }
 
 //TODO: add unit test for each implementation of network plugin interface

